lib/mayaUsd/render/vp2ShaderFragments/shaderFragments.cpp

```diff
@@ -330,6 +344,121 @@ MStatus HdVP2ShaderFragments::registerFragments()
         }
     }
 
+    {
+        // Register UVTexture readers for some common working spaces:
+        std::string uvTextureFile = TfStringPrintf("%s.xml", _tokens->UsdUVTexture.GetText());
+        uvTextureFile = _GetResourcePath(uvTextureFile);
+        std::ifstream xmlFile(uvTextureFile.c_str());
+        std::string   xmlString;
+        xmlFile.seekg(0, std::ios::end);
+        xmlString.reserve(xmlFile.tellg());
+        xmlFile.seekg(0, std::ios::beg);
+        xmlString.assign(
+            (std::istreambuf_iterator<char>(xmlFile)), std::istreambuf_iterator<char>());
+
+        const std::regex RE_FRAG("UsdUVTexture");
+        const std::regex RE_GLSL("TO_MAYA_COLOR_SPACE_GLSL");
+        const std::regex RE_HLSL("TO_MAYA_COLOR_SPACE_HLSL");
+        const std::regex RE_CG("TO_MAYA_COLOR_SPACE_CG");
+
+        // We create custom UsdUVTexture fragments. We use the original UsdUVTexture.xml as template
+        // where we replace four important tokens with custom code:
+        //
+        //  - UsdUVTexture gets renamed by appending the name of the working space we are targetting
+        //  - The three TO_MAYA_COLOR_SPACE_* markers gets replaced by the correct 4x4 matrix
+        //  multiplication that will take the color from a "scene-linear Rec 709/sRGB" to the final
+        //  color space. The syntax of the multiplication varies by shading language, and the
+        //  template is:
+        //     GLSL: outColor = mat4( ... ) * outColor;
+        //     HLSL: outColor = mul(outColor, float4x4( ... ));
+        //       CG: outColor = mul(float4x4( transpose( ... ) ), outColor);
+        //    Where the ... denotes a 4x4 matrix expanded from the 3x3 matrix passed in as parameter
+        //    to the function (alpha is always left untouched).
+        //
+        auto registerSpace = [&](const std::string& spaceName,
+                                 const std::string& fragName,
+                                 const float*       convMatrix) {
+            if (!fragmentManager->hasFragment(fragName.c_str())) {
+                std::string xmlFinal = std::regex_replace(xmlString, RE_FRAG, fragName.c_str());
+                if (convMatrix) {
+                    std::stringstream op_glsl, op_hlsl, op_cg;
+                    op_glsl << "outColor = mat4(";
+                    op_hlsl << "outColor = mul(outColor, float4x4(";
+                    op_cg << "outColor = mul(float4x4(";
```

Comment:
I have some concerns about the matrix implementation which is different from the OpenColorIO one (refer to 
https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/master/src/OpenColorIO/GpuShaderUtils.cpp) where the transpose is done for HLSL only!

When I did some search on the Matrix access for GLSL, HLSL and Cg, I find that HLSL is column first, and GLSL & Cg are raw first!

Reply:
Yes, this is a very complex issue. You need also to consider that the OpenGL code does mat4 * vec4 while the HLSL code does vec4 * mat4.
When in doubt, the best option is to see what code OCIO generates and replicate that exactly, so yes, I do admit I cheated and ran Maya thru a debugger to see the exact shader code it generated for each shading language and replicated it exactly (with an eye for transposed data).
To test: use the most asymmetric transform "scene-linear DCI-P3 D65" and see what happens in Maya. The color will either be correct, or completely off.

Reply:
Good point. You can copy & paste OCIO for now. You already have the link in the comment.